File: detect_left_recursion.java
import java.util.*;

public class LeftRecursion {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter production (like A->Aa|b): ");
        String prod = sc.next();

        char lhs = prod.charAt(0);
        int arrowPos = prod.indexOf("->");
        if (arrowPos == -1) {
            System.out.println("Invalid production format.");
            return;
        }

        String rhs = prod.substring(arrowPos + 2);
        String[] alts = rhs.split("\\|");
        boolean leftRec = false;

        for (String alt : alts) {
            alt = alt.trim();
            if (!alt.isEmpty() && alt.charAt(0) == lhs) {
                leftRec = true;
                break;
            }
        }

        if (leftRec)
            System.out.println("Immediate left recursion detected for " + lhs + ".");
        else
            System.out.println("No immediate left recursion detected for " + lhs + ".");
    }
}

/*
Example Inputs & Outputs:

1️⃣ Input: A->Aa|b  
    Output: Immediate left recursion detected for A.

2️⃣ Input: E->E+T|T  
    Output: Immediate left recursion detected for E.

3️⃣ Input: S->aS|b  
    Output: No immediate left recursion detected for S.
*/

File: detect_left_recursion.c
#include <stdio.h>
#include <string.h>

int main() {
    char prod[200];
    printf("Enter production (like A->Aa|b): ");
    scanf("%199s", prod);

    // find LHS nonterminal (before "->")
    char lhs = prod[0];

    // find start of RHS (after "->")
    char *rhs = strstr(prod, "->");
    if (!rhs) {
        printf("Invalid production format.\n");
        return 0;
    }
    rhs += 2; // move past "->"

    // split alternatives by '|', check each alternative's first symbol
    int leftRec = 0;
    char *alt = strtok(rhs, "|");
    while (alt != NULL) {
        // skip leading spaces (if any)
        int i = 0;
        while (alt[i] == ' ' || alt[i] == '\t') i++;

        if (alt[i] == lhs) { // immediate left recursion detected
            leftRec = 1;
            break;
        }
        alt = strtok(NULL, "|");
    }

    if (leftRec)
        printf("Immediate left recursion detected for %c.\n", lhs);
    else
        printf("No immediate left recursion detected for %c.\n", lhs);

    return 0;
}

/*
Input: A->Aa|b
Output: Immediate left recursion detected for A.

Input: E->E+T|T
Output: Immediate left recursion detected for E.

Input: S->aS|b
Output: No immediate left recursion detected for S.
*/